util/retry/retry.go

```diff
@@ -44,81 +37,94 @@ func (re *MaxAttemptsError) Error() string {
 	return fmt.Sprintf("maximum number of attempts exceeded %d", re.MaxAttempts)
 }
 
-const (
-	// Break indicates the retry loop is finished and should return
-	// the result of the retry worker function.
-	Break Status = iota
-	// Reset indicates that the retry loop should be reset with
-	// no backoff for an immediate retry.
-	Reset
-	// Continue indicates that the retry loop should continue with
-	// another iteration of backoff / retry.
-	Continue
-)
+// Retry implements the methods which control execution of retry loops. See
+// Reset, Stop below.
+type Retry struct {
+	// TODO(tamird): use backoff.BackOff here (it's an interface)
+	backOff backoff.ExponentialBackOff
+	stopped bool // Used for explicit stopping
+	count   uint
+}
+
+// Reset resets the exponential backoff and causes an immediate retry.
+func (r *Retry) Reset() {
+	r.count = 0
+	r.backOff.Reset()
+}
+
+// Stop causes the retry loop to stop on the next iteration.
+func (r *Retry) Stop() {
+	r.stopped = true
+}
 
 // Options provides control of retry loop logic via the
 // WithBackoffOptions method.
 type Options struct {
+	// TODO(tamird): use backoff.BackOff here (it's an interface)
+	BackOff     backoff.ExponentialBackOff
 	Tag         string        // Tag for helpful logging of backoffs
-	Backoff     time.Duration // Default retry backoff interval
-	MaxBackoff  time.Duration // Maximum retry backoff interval
-	Constant    float64       // Default backoff constant
-	MaxAttempts int           // Maximum number of attempts (0 for infinite)
+	MaxAttempts uint          // Maximum number of attempts (0 for infinite)
 	UseV1Info   bool          // Use verbose V(1) level for log messages
 	Stopper     *util.Stopper // Optionally end retry loop on stopper signal
 }
 
-// WithBackoff implements retry with exponential backoff using
-// the supplied options as parameters. When fn returns Continue
-// and the number of retry attempts haven't been exhausted, fn is
-// retried. When fn returns Break, retry ends. As a special case,
-// if fn returns Reset, the backoff and retry count are reset to
-// starting values and the next retry occurs immediately. Returns an
-// error if the maximum number of retries is exceeded or if the fn
-// returns an error.
-func WithBackoff(opts Options, fn func() (Status, error)) error {
-	backoff := opts.Backoff
+// WithBackoff implements retry with exponential backoff using the supplied
+// options as parameters. When fn returns non-nil and the number of retry
+// attempts haven't been exhausted, fn is retried. When fn returns nil or
+// Stop() is called, retry ends. As a special case, if Reset is called, the
+// backoff and retry count are reset to starting values and the next retry
+// occurs immediately. Returns an error if the maximum number of retries is
+// exceeded or if fn returns an error on the last attempt.
+func WithBackoff(opts Options, fn func(*Retry) error) error {
```

Comment:
I think the `for` loop approach is probably the most idiomatic in Go. I'll whip something up.